### PR TITLE
lib: add a UsbBus::force_reset implementation

### PR DESCRIPTION
**Pull request by rmsyn**
Implements the `force_reset` function for `usb_device::UsbBus`, that simply sets the `UDCON::detach` bit.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -463,6 +463,17 @@ impl usb_device::bus::UsbBus for UsbBus {
...
         })
     }
+
+    fn force_reset(&self) -> usb_device::Result<()> {
+        interrupt::free(|cs| {
+            self.usb
+                .borrow(cs)
+                .udcon
+                .modify(|_, w| w.detach().set_bit());
+
```
Comment (agausmann) [inline src/lib.rs:479]: Should this also clear the detach bit? From the datasheet:

> It is possible to re-enumerate a device, simply by setting and clearing the DETACH bit (but firmware must take
in account a debouncing delay of some milliseconds).

The only place where the detach bit is cleared in this driver is in `enable()`. However the [trait docs](https://docs.rs/usb-device/latest/usb_device/bus/trait.UsbBus.html#method.force_reset) don't specify any steps to re-initialize the device after calling `force_reset()`, so I think this should be a self-contained detach and attach sequence.

Comment (agausmann) [reply src/lib.rs:479]: FWIW, [that's what the nRF USB driver does](https://docs.rs/nrf-usbd/0.2.0/src/nrf_usbd/usbd.rs.html#733-747)

Comment (agausmann) [reply src/lib.rs:479]: I'll probably take care of this myself. If we don't want to depend on `avr_hal::Delay`, then we need to write a cycle delay routine (like what the nRF driver has) to delay for at least 1 millisecond (16000 cycles at 16 MHz)

For now I'll put it in this crate but I'll also submit a pull request to `avr-device`. I think that would be useful for them to have.

Comment (rmsyn) [reply src/lib.rs:479]: > I'll probably take care of this myself. If we don't want to depend on avr_hal::Delay, then we need to write a cycle delay routine (like what the nRF driver has) to delay for at least 1 millisecond (16000 cycles at 16 MHz)

Sounds good, I mostly need this for a project I'm working on, and the upstream for that project just does this simple `detach`. 

> The only place where the detach bit is cleared in this driver is in enable(). However the [trait docs](https://docs.rs/usb-device/latest/usb_device/bus/trait.UsbBus.html#method.force_reset) don't specify any steps to re-initialize the device after calling force_reset(), so I think this should be a self-contained detach and attach sequence.

AFAICT, the `enable` call is only made during `UsbDevice` creation, since the rest of the trait API looks to intentionally make it impossible to directly call `enable`.

There does seem to be some verbiage about the host resetting all device state after a reset, so maybe it is intended for users to simply re-create a `UsbDevice` after a `force_reset`, IDK

**Comment by agausmann** [comment]
@rmsyn Are you able to test this change?

Since I don't have permission to push to your branch, I updated it here: [`agausmann:feature/detach`](https://github.com/agausmann/atmega-usbd/tree/feature/detach) 875493126cc04e1dd229a83085de934f4660c18c

**Comment by rmsyn** [comment]
> Are you able to test this change?

I'm getting pretty close to being able to run the firmware on real hardware. I'll use your changes, and provide updates on how it works.

FWIW, it looks good to me, and passes the build test.

> Since I don't have permission to push to your branch, I updated it here: [agausmann:feature/detach](https://github.com/agausmann/atmega-usbd/tree/feature/detach) https://github.com/agausmann/atmega-usbd/commit/875493126cc04e1dd229a83085de934f4660c18c

I changed to allow edits from maintainers, and cherry-picked your changes. Let me know if you want me to drop the cherry-pick, so you can add it yourself.

**Comment by agausmann** [comment]
Alright, the `delay_cycles` method will be added to `avr-device` in its next release

**Comment by agausmann** [comment]
I actually haven't tested the delay yet. :sweat_smile:  I will write a blinky using it later today. Although the generated assembly looks good

**Comment by agausmann** [comment]
Merged as bdbc13dfdb1300565645668277ac32a7a6914d1f
